### PR TITLE
Fix init order of SetupWithClassicDesktopLifetime

### LINKED ISSUE #20683: 'Quit' option in dock menu on macOS can't be canceled when `SetupWithClassicDesktopLifetime()` is used
### Describe the bug

When an app is set up with `ClassicDesktopStyleApplicationLifetimeExtensions.SetupWithClassicDesktopLifetime()`, canceling shutdown by setting `ShutdownRequestedEventArgs.Cancel` to `true` in an `IClassicDesktopStyleApplicationLifetime.ShutdownRequested` event handler does not cancel shutdown when the user selects to quit the app by right-clicking the app's icon in the macOS dock then selecting the 'Quit' item from the menu.

Image

Canceling shutdown does work when the user closes the window with the red window button, and when the user quits the app with the macOS menu bar.

### To Reproduce

1. Set up an app with `ClassicDesktopStyleApplicationLifetimeExtensions.SetupWithClassicDesktopLifetime()`
2. Add an `IClassicDesktopStyleApplicationLifetime.ShutdownRequested` handler that sets `ShutdownRequestedEventArgs.Cancel` to `true`
3. Run the app on macOS
4. Right-click the app's icon in the dock
5. Select the 'Quit' menu item
6. The `IClassicDesktopStyleApplicationLifetime.ShutdownRequested` handler is called twice, but canceling shutdown does not work

- [Repo which demonstrates bug.](https://github.com/jtcyacomb/AvaloniaShutdownBug/commits/setup/)
- [Branch which uses `ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime()` which does not cause the same bug.](https://github.com/jtcyacomb/AvaloniaShutdownBug/commits/start)

### Expected behavior

Canceling shutdown should work, as it does with `ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime()`.

### Avalonia version

11.3.12

### OS

macOS

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by MrJul**
## What does the pull request do?
This PR fixes an initialization order bug when `AppBuilder.SetupWithClassicDesktopLifetime()` is called instead of `StartWithClassicDesktopLifetime()`.

## What is the current behavior?
When `SetupWithClassicDesktopLifetime` is used, internal events are registered before the `AppBuilder` initializers run: at that point, many subsystems aren't available yet, causing issues such as `ShutdownRequested` not being hooked up. Additionally, `Startup` is called way earlier, or might be called twice if `Start()` is called manually on the created lifetime.

A related problem arises when calling `AppBuilder.SetupWithLifetime()` manually without `Start`: the internal event handlers aren't even initialized.

On macOS, this causes the `ShutdownRequested` event not to be called when quitting the app from the dock, then triggering a double close of the main window (and hitting a debug assertion).

## What is the updated/expected behavior with this PR?
`StartWithClassicDesktopLifetime`, `SetupWithClassicDesktopLifetime` and `SetupWithLifetime` initialize the subsystems and event handlers in the same exact order. 

Quitting a macOS application raises the `ShutdownRequested` event correctly and respects its cancellation.

Unit tests have been added.

## How was the solution implemented (if it's not obvious)?
An internal `ISetupApplicationLifetime` is added and implemented by `ClassicDesktopStyleApplicationLifetime`. The `AppBuilder` is responsible for calling it to initialize the lifetime in the correct order properly.

## Fixed issues
 - Fixes #20683

### CODE REVIEW

File: src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs [hunk 1]
```diff
@@ -93,16 +96,33 @@ internal void SubscribeGlobalEvents()
...
+            _platformLifetimeEventsImpl = AvaloniaLocator.Current.GetService<IPlatformLifetimeEventsImpl>();
+            _platformLifetimeEventsImpl?.ShutdownRequested += OnShutdownRequested;
+        }
 
-            if (lifetimeEvents != null)
-                lifetimeEvents.ShutdownRequested += OnShutdownRequested;
+        void ISetupApplicationLifetime.AfterAppInit()
+        {
+            AfterInit();
+            InvokeStartup(Args);
```
Comment (jsuarezruiz) [inline src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs]: Calling the public `SetupWithLifetime(new ClassicDesktopStyleApplicationLifetime())` path reaches this with `Args == null`. This raises `Startup` with an empty array and sets `_startupCalled`, so a handler registered after setup is never invoked by `Start()`, and `Start(["foo", "bar"])` cannot deliver its actual arguments.

I did some quick tests with the same sequence against the exact base and PR commits: base invokes the post-setup subscriber once with `["foo", "bar"]`; this PR invokes it zero times, while a pre-setup subscriber receives `[]`.

Could setup-time `Startup` be conditional on arguments already being available?

```csharp
AfterInit();
if (Args is not null)
    InvokeStartup(Args);
```

`SetupWithClassicDesktopLifetime` sets `Args` before setup, so this preserves its historical no-`Start()` behavior while leaving `StartCore(args)` as the fallback for manual `SetupWithLifetime`.

**Review by jsuarezruiz — CHANGES_REQUESTED** [review]

File: src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs (hunk 1 above)
Comment (MrJul) [reply src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs]: That's a compromise I explicitly documented here:  https://github.com/AvaloniaUI/Avalonia/pull/21954/changes#diff-3fc2a19a07871f16629e9f5e591608ca565a74197e6f0d2a7b57cf191c80a8dfR164-R167

> Ideally, the Startup event should be invoked only when Start() is called. However, it was historically invoked in SetupWithClassicDesktopLifetime to account for programs that don't call Start at all, so let's keep this behavior.
> Technically, the args could be different between Setup and Start. In practice, they won't for all callers we control.

As you mentioned, `Args` must be set before setup.

>```csharp
> if (Args is not null)
>     InvokeStartup(Args);
> ```

Doing this means `Startup` will never be called if `Setup` is called without `Start` (and no `Args`). `Startup` _should_ be invoked only on `Start()` in my opinion, but that's not what happened previously and changing this in the v12.x timeframe would be a breaking change.

Comment (jsuarezruiz) [reply src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs]: Thanks for the detailed answer. I agree that we should preserve the existing `SetupWithClassicDesktopLifetime` behavior now. The null guard still does that because this helper sets `Args` before setup.
The no-arguments case is different. When using the public `SetupWithLifetime(new ClassicDesktopStyleApplicationLifetime())` path, on `master` branch, that path does not raise `Startup` during setup, so `Start(["param1", "param2"])` raises it later with the actual arguments. On this PR, setup consumes the event with an empty array, so a handler registered afterward is never called.
Could we add a related test and keep the null check?

```
[Fact]
public void SetupWithLifetime_Should_Defer_Startup_Until_Start()
{
    using var lifetime = new ClassicDesktopStyleApplicationLifetime();
    string[]? startupArgs = null;
    var raised = 0;

    CreateAppBuilder().SetupWithLifetime(lifetime);

    lifetime.Startup += (_, e) =>
    {
        ++raised;
        startupArgs = e.Args;
    };

    Dispatcher.UIThread.Post(Dispatcher.UIThread.ExitAllFrames);
    lifetime.Start(["param1", "param2"]);

    Assert.Equal(1, raised);
    Assert.Equal(["param1", "param2"], startupArgs!);
}
```

**Review by jsuarezruiz — DISMISSED** [review]

**Comment by MrJul** [comment]
~I disagree. The whole point of this PR is to make it consistent: `Startup` is always raised from the same place, here `Setup` (unfortunately). There might be no `Start()` call at all!~

~In other words, `Setup(new ClassicDesktopStyleApplicationLifetime())` and `SetupWithClassicDesktopLifetime` should have the same behavior with regard to when/if `Startup` is raised. That's why this PR adds it directly to the base builder instead of special-casing just one path.~

~In v13, we should remove this hack and take the breaking change that `Startup` is only called when `Start` is, solving both issues and respecting the arguments.~

**Comment by MrJul** [comment]
Having thought about it more, the fact that `Setup` was raising `Startup` is also part of the bug. The whole initialization order was incorrect. I've changed my stance: `Startup` is now raised only when `Start` is called, not on `Setup`.

This makes the whole thing sane, without hacks.

The following numbers helped this decision:
 - [SetupWithClassicDesktopLifetime](https://github.com/search?q=SetupWithClassicDesktopLifetime+NOT+is%3Afork&type=code): 78 hits
 -  [StartWithClassicDesktopLifetime](https://github.com/search?q=StartWithClassicDesktopLifetime+NOT+is%3Afork&type=code) : 19800 hits

Users of `Setup` are few, and such users know when their app starts, so the `Startup` event is irrelevant there, especially since it was called _immediately_ from `Setup`, which is user-controlled.

**Review by kekekeks — APPROVED** [review]
